presentation/src/main/java/com/anytypeio/anytype/presentation/history/VersionHistoryViewModel.kt

```diff
@@ -180,39 +217,61 @@ class VersionHistoryViewModel(
             ).process(
                 failure = {
                     Timber.e(it, "Error while fetching new member")
-                    _viewState.value = VersionHistoryState.Error.SpaceMembers(it.message.orEmpty())
+                    _viewState.value = VersionHistoryState.Error.SpaceMembers
                 },
                 success = { members ->
                     if (members.isEmpty()) {
                         _viewState.value =
-                            VersionHistoryState.Error.SpaceMembers("No members found")
+                            VersionHistoryState.Error.SpaceMembers
                     } else {
-                        getHistoryVersions(vmParams.objectId, members)
+                        _members.value = members
                     }
                 }
             )
         }
     }
 
-    private fun getHistoryVersions(objectId: String, members: List<ObjectWrapper.Basic>) {
+    private fun getHistoryVersions(objectId: String, latestVersionId: String = "") {
         viewModelScope.launch {
-            val params = GetVersions.Params(objectId = objectId)
+            val params = GetVersions.Params(
+                objectId = objectId,
+                lastVersion = latestVersionId,
+                limit = VERSIONS_MAX_LIMIT
+            )
             getVersions.async(params).fold(
                 onSuccess = { versions ->
-                    if (versions.isEmpty()) {
-                        _viewState.value = VersionHistoryState.Error.NoVersions
+                    canPaginate = versions.size == VERSIONS_MAX_LIMIT
+                    if (latestVersionId.isEmpty()) {
+                        if (versions.isEmpty()) {
+                            _viewState.value = VersionHistoryState.Error.NoVersions
+                        } else {
+                            _list.value = versions
+                        }
                     } else {
-                        handleVersionsSuccess(versions, members)
+                        _list.value += versions
+                    }
+                    listState = ListState.IDLE
+                    if (canPaginate) {
+                        latestVisibleVersionId = versions.lastOrNull()?.id ?: ""
                     }
                 },
                 onFailure = {
-                    _viewState.value = VersionHistoryState.Error.GetVersions(it.message.orEmpty())
+                    _viewState.value = VersionHistoryState.Error.GetVersions
+                    listState =
+                        if (latestVersionId.isEmpty()) ListState.ERROR else ListState.PAGINATION_EXHAUST
                 },
                 onLoading = {}
             )
         }
     }
 
+    override fun onCleared() {
```

Comment:
do we need this behavior, since this vm will be destroyed after this method call?

Reply:
deleted